File: backend/app/modules/patient_records/models/patient_record.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING
from uuid import uuid4

from sqlalchemy import DateTime, ForeignKey, Index, Text, Enum, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship,
)

from app.database.base import Base
from app.modules.patient_records.enums import RecordStatus

if TYPE_CHECKING:
    from .diagnosis import PatientRecordDiagnosis
    from .prescription import PatientRecordPrescription
    from .attachment import PatientRecordAttachment
    from .followup import PatientRecordFollowup
    from .audit_log import PatientRecordAuditLog


class PatientRecord(Base):
    """
    Stores the complete clinical record for a patient.

    An optional ``appointment_id`` links the record to an appointment
    when one exists. Records may also be created without an appointment
    (walk-in notes, historical entry, etc.).
    """

    __tablename__ = "patient_records"

    id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    patient_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("patients.id"),
        nullable=False,
        index=True,
    )

    appointment_id: Mapped[UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("appointments.id"),
        nullable=True,
        unique=True,
        index=True,
    )

    status: Mapped[RecordStatus] = mapped_column(
        Enum(RecordStatus),
        nullable=False,
        default=RecordStatus.DRAFT,
    )

    chief_complaint: Mapped[str | None] = mapped_column(Text)
    clinical_notes: Mapped[str | None] = mapped_column(Text)
    doctor_remarks: Mapped[str | None] = mapped_column(Text)
    treatment_recommendation: Mapped[str | None] = mapped_column(Text)

    systemic_diseases: Mapped[str | None] = mapped_column(Text)
    surgeries: Mapped[str | None] = mapped_column(Text)
    medications: Mapped[str | None] = mapped_column(Text)
    habits: Mapped[str | None] = mapped_column(Text)
    medical_alerts: Mapped[str | None] = mapped_column(Text)

    allergies: Mapped[str | None] = mapped_column(Text)
    dental_history: Mapped[str | None] = mapped_column(Text)

    is_finalized: Mapped[bool] = mapped_column(
        default=False,
        nullable=False,
    )

    is_deleted: Mapped[bool] = mapped_column(
        default=False,
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    diagnoses: Mapped[list["PatientRecordDiagnosis"]] = relationship(
        back_populates="patient_record",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    prescriptions: Mapped[list["PatientRecordPrescription"]] = relationship(
        "PatientRecordPrescription",
        back_populates="patient_record",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    attachments: Mapped[list["PatientRecordAttachment"]] = relationship(
        back_populates="patient_record",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    followups: Mapped[list["PatientRecordFollowup"]] = relationship(
        back_populates="patient_record",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    audit_logs: Mapped[list["PatientRecordAuditLog"]] = relationship(
        back_populates="patient_record",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    @property
    def diagnosis_count(self) -> int:
        return len(self.diagnoses) if self.diagnoses else 0

    @property
    def prescription_count(self) -> int:
        return len(self.prescriptions) if self.prescriptions else 0

    @property
    def attachment_count(self) -> int:
        return len(self.attachments) if self.attachments else 0

    @property
    def followup_count(self) -> int:
        return len(self.followups) if self.followups else 0

    __table_args__ = (
        Index("ix_patient_records_status", "status"),
        Index("ix_patient_records_is_deleted", "is_deleted"),
    )

File: frontend/src/types/patientRecord.ts
/* ============================================================
 * Patient Record Types
 *
 * Strictly mirrors backend `app/modules/patient_records/`:
 * - enums/  -> RecordStatus, DiagnosisType, AttachmentType
 * - schemas -> request + response DTOs (snake_case, verbatim)
 * - pagination -> {items, total, page, page_size, pages} — the
 *   patient-records module uses `pages` (NOT `total_pages`, which
 *   is the treatment-plan shape).
 *
 * Backend enums are uppercase string-valued, so string-literal
 * unions are used (the project's `erasableSyntaxOnly` tsconfig
 * forbids TS enums).
 * ============================================================ */

/* ── Enums (string-literal unions) ──────────────────────────────── */

/** Backend `RecordStatus` (enums/record_status.py). Sent/received verbatim. */
export type RecordStatus =
  | 'DRAFT'
  | 'IN_PROGRESS'
  | 'UNDER_REVIEW'
  | 'COMPLETED'
  | 'FINALIZED'
  | 'LOCKED';

/** Backend `DiagnosisType` (enums/diagnosis_type.py). */
export type DiagnosisType = 'PROVISIONAL' | 'CONFIRMED';

/** Backend `AttachmentType` (enums/attachment_type.py) — no VIDEO/OTHER. */
export type AttachmentType = 'IMAGE' | 'PDF' | 'REPORT' | 'SCAN' | 'DOCUMENT';

/* ── Pagination (module-specific shape) ─────────────────────────── */

/** Paginated envelope returned by every patient-records list endpoint. */
export interface PatientRecordListEnvelope<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  /** `ceil(total/page_size)`; 0 when total = 0. */
  pages: number;
}

/* ── Record list / detail DTOs ──────────────────────────────────── */

/**
 * Row returned by `GET /patient-records` and `GET /patient-records/patient/{id}`.
 * No patient/appointment names — resolved client-side (backend returns ids only).
 */
export interface PatientRecordListItem {
  id: string;
  patient_id: string;
  appointment_id: string;
  status: RecordStatus;
  is_finalized: boolean;
  chief_complaint: string | null;
  created_at: string;
}

/** Lightweight nested child shapes embedded in the record-detail aggregate. */
export interface DiagnosisNestedResponse {
  id: string;
  diagnosis_name: string;
  diagnosis_type: DiagnosisType;
}

export interface PrescriptionItemNestedResponse {
  id: string;
  medicine_name: string;
  dosage: string;
  frequency: string;
  duration: string;
}

export interface PrescriptionNestedResponse {
  id: string;
  prescribed_at: string;
  items: PrescriptionItemNestedResponse[];
}

export interface FollowupNestedResponse {
  id: string;
  followup_date: string;
  notes: string | null;
}

export interface AttachmentNestedResponse {
  id: string;
  attachment_type: AttachmentType;
  file_name: string;
  mime_type: string | null;
}

/**
 * Audit entry embedded in the record-detail response — visible to ALL
 * read roles here (O4); standalone audit endpoints are admin-only.
 */
export interface AuditNestedResponse {
  id: string;
  action: string;
  performed_by: number;
  performed_at: string;
}

/**
 * Full record aggregate (`PatientRecordResponse` schema) — returned by
 * detail, by-appointment and every record mutation. Carries all 11
 * clinical/medical text fields, all nested child families, the embedded
 * audit log and backend-computed counts. Counts are displayed as-is —
 * never recomputed client-side.
 */
export interface PatientRecordResponse {
  id: string;
  patient_id: string;
  appointment_id: string;
  status: RecordStatus;
  is_finalized: boolean;
  created_at: string;
  updated_at: string;
  chief_complaint: string | null;
  clinical_notes: string | null;
  doctor_remarks: string | null;
  treatment_recommendation: string | null;
  systemic_diseases: string | null;
  surgeries: string | null;
  medications: string | null;
  habits: string | null;
  medical_alerts: string | null;
  allergies: string | null;
  dental_history: string | null;
  diagnoses: DiagnosisNestedResponse[];
  prescriptions: PrescriptionNestedResponse[];
  followups: FollowupNestedResponse[];
  attachments: AttachmentNestedResponse[];
  audit_logs: AuditNestedResponse[];
  diagnosis_count: number;
  prescription_count: number;
  attachment_count: number;
  followup_count: number;
}

/* ── Request DTOs (mirror schemas, extra="forbid") ──────────────── */

/**
 * POST /patient-records payload. `patient_id`/`appointment_id` required;
 * the 11 text fields optional (≤ limits) and stripped — empty → null.
 */
export interface PatientRecordCreateRequest {
  patient_id: string;
  appointment_id: string;
  chief_complaint?: string | null;
  clinical_notes?: string | null;
  doctor_remarks?: string | null;
  treatment_recommendation?: string | null;
  systemic_diseases?: string | null;
  surgeries?: string | null;
  medications?: string | null;
  habits?: string | null;
  medical_alerts?: string | null;
  allergies?: string | null;
  dental_history?: string | null;
}

/**
 * PATCH /patient-records/{id} payload — all fields optional
 * (`exclude_unset`): omitted keys untouched, explicit `null` clears.
 */
export type PatientRecordUpdateRequest = Partial<PatientRecordCreateRequest>;

/** POST /patient-records/{id}/finalize — must be the literal `true`. */
export interface PatientRecordFinalizeRequest {
  confirm: true;
}

/* ── Diagnoses ──────────────────────────────────────────────────── */

export interface DiagnosisCreateRequest {
  /** 2–255 chars, non-empty after strip */
  diagnosis_name: string;
  diagnosis_type: DiagnosisType;
  /** ≤ 2000 */
  notes?: string | null;
}

export interface DiagnosisUpdateRequest {
  diagnosis_name?: string;
  diagnosis_type?: DiagnosisType;
  notes?: string | null;
}

/** Full diagnosis response (also the item-router response). */
export interface DiagnosisResponse {
  id: string;
  patient_record_id: string;
  diagnosis_name: string;
  diagnosis_type: DiagnosisType;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

/** Row from `GET /patient-records/{id}/diagnoses`. */
export interface DiagnosisListItem {
  id: string;
  diagnosis_name: string;
  diagnosis_type: DiagnosisType;
  created_at: string;
}

/* ── Prescriptions ──────────────────────────────────────────────── */

export interface PrescriptionItemCreateRequest {
  /** 2–255 */
  medicine_name: string;
  /** 1–100 */
  dosage: string;
  /** 1–100 */
  frequency: string;
  /** 1–100 */
  duration: string;
  /** ≤ 2000 */
  instructions?: string | null;
}

export interface PrescriptionItemUpdateRequest {
  medicine_name?: string;
  dosage?: string;
  frequency?: string;
  duration?: string;
  instructions?: string | null;
}

/** POST /patient-records/{id}/prescriptions — items 1–20, atomic. */
export interface PrescriptionCreateRequest {
  /** ≤ 3000 */
  notes?: string | null;
  items: PrescriptionItemCreateRequest[];
}

/** PATCH /prescriptions/{id} — NOTES ONLY (items are managed separately). */
export interface PrescriptionUpdateRequest {
  notes?: string | null;
}

/** Row from `GET /patient-records/{id}/prescriptions`. */
export interface PrescriptionListItem {
  id: string;
  prescribed_at: string;
  prescribed_by: number;
  medicine_count: number;
}

/** Full prescription incl. items (`PrescriptionResponse` schema). */
export interface PrescriptionResponse {
  id: string;
  patient_record_id: string;
  prescribed_by: number;
  prescribed_at: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
  items: PrescriptionItemResponse[];
}

export interface PrescriptionItemResponse {
  id: string;
  prescription_id: string;
  medicine_name: string;
  dosage: string;
  frequency: string;
  duration: string;
  instructions: string | null;
  created_at: string;
  updated_at: string;
}

/* ── Attachments (metadata only — no upload/download) ───────────── */

export interface AttachmentCreateRequest {
  attachment_type: AttachmentType;
  /** 1–255, non-empty */
  file_name: string;
  /** 1–1000 — client-supplied path/link string (no file upload) */
  file_path: string;
  /** ≤ 100 */
  mime_type?: string | null;
  /** ≥ 0; backend rejects > 50 MB (400) */
  file_size?: number | null;
}

export interface AttachmentUpdateRequest {
  attachment_type?: AttachmentType;
  file_name?: string;
  mime_type?: string | null;
  file_size?: number | null;
  /** file_path is IMMUTABLE — never accepted by the update endpoint. */
}

/** Row from `GET /patient-records/{id}/attachments`. */
export interface AttachmentListItem {
  id: string;
  attachment_type: AttachmentType;
  file_name: string;
  mime_type: string | null;
  file_size: number | null;
  created_at: string;
}

/* ── Follow-ups ─────────────────────────────────────────────────── */

export interface FollowupCreateRequest {
  /** YYYY-MM-DD — must be today or future (else 400 business rule) */
  followup_date: string;
  /** ≤ 2000 */
  notes?: string | null;
}

export interface FollowupUpdateRequest {
  followup_date?: string;
  notes?: string | null;
}

/** Row from `GET /patient-records/{id}/followups` (followup_date ASC). */
export interface FollowupListItem {
  id: string;
  followup_date: string;
  notes: string | null;
  created_at: string;
}

/* ── Query params ───────────────────────────────────────────────── */

/** Query params for GET /patient-records (no sort params exist). */
export interface PatientRecordListParams {
  /** Matches chief_complaint OR clinical_notes (substring). */
  search?: string;
  status?: RecordStatus;
  is_finalized?: boolean;
  patient_id?: string;
  page?: number;
  page_size?: number;
}

/** Query params for the paginated child lists. */
export interface ChildListParams {
  page?: number;
  page_size?: number;
}

/* ── UI form values (never sent to the API as-is) ──────────────── */

/** Create/edit record form model (presentational). */
export interface PatientRecordFormValues {
  patient_id: string;
  appointment_id: string;
  chief_complaint: string;
  clinical_notes: string;
  doctor_remarks: string;
  treatment_recommendation: string;
  systemic_diseases: string;
  surgeries: string;
  medications: string;
  habits: string;
  medical_alerts: string;
  allergies: string;
  dental_history: string;
}

export interface DiagnosisFormValues {
  diagnosis_name: string;
  diagnosis_type: string;
  notes: string;
}

export interface PrescriptionItemFormValues {
  medicine_name: string;
  dosage: string;
  frequency: string;
  duration: string;
  instructions: string;
}

export interface PrescriptionFormValues {
  notes: string;
  items: PrescriptionItemFormValues[];
}

export interface AttachmentFormValues {
  attachment_type: string;
  file_name: string;
  file_path: string;
  mime_type: string;
  file_size: string;
}

export interface FollowupFormValues {
  followup_date: string;
  notes: string;
}

/* ── Enriched display types ─────────────────────────────────────── */

/** Record row enriched with resolved display names (backend returns ids only). */
export interface EnrichedPatientRecord extends PatientRecordListItem {
  patient_name: string | null;
  appointment_number: string | null;
}
